File: app/src/main/java/ru/melod1n/schedule/activity/ThemesActivity.java
package ru.melod1n.schedule.activity;

import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import ru.melod1n.schedule.R;
import ru.melod1n.schedule.adapter.ThemeAdapter;
import ru.melod1n.schedule.common.ThemeEngine;
import ru.melod1n.schedule.current.BaseActivity;
import ru.melod1n.schedule.items.ThemeItem;
import ru.melod1n.schedule.widget.Toolbar;

public class ThemesActivity extends BaseActivity {

    public static final int REQUEST_PICK_THEME = 0;
    public static final int REQUEST_PICK_DAY_THEME = 1;
    public static final int REQUEST_PICK_NIGHT_THEME = 2;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.recyclerView)
    RecyclerView list;

    @BindView(R.id.refresh)
    SwipeRefreshLayout refreshLayout;

    private ThemeAdapter adapter;
    private int request;

    private volatile boolean animating;
    private Drawable navigationIcon;

    private View rootView;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_themes);
        ButterKnife.bind(this);

        request = getIntent().getIntExtra("request", -1);

        applyBackground();
        applyTitle();

        rootView = toolbar.getRootView();

        navigationIcon = ContextCompat.getDrawable(this, R.drawable.ic_arrow_backward);

        toolbar.setNavigationOnClickListener(view -> onBackPressed());

        refreshLayout.setOnRefreshListener(this::onRefresh);

        list.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        getThemes();

        if (savedInstanceState == null) {
            scrollToCurrentTheme();
        }
    }

    private void scrollToCurrentTheme() {
        int index = 0;

        for (int i = 0; i < adapter.getItemCount(); i++) {
            ThemeItem item = adapter.getItem(i);

            if (item.equals(ThemeEngine.getCurrentTheme())) index = i;
        }

        list.smoothScrollToPosition(index);
    }

    private void applyTitle() {
        ThemeItem theme = ThemeEngine.getCurrentTheme();
        ThemeItem dayTheme = ThemeEngine.getDayTheme();
        ThemeItem nightTheme = ThemeEngine.getNightTheme();

        String formatString = "%s – %s";

        String title;

        int subtitle;

        switch (request) {
            default:
            case REQUEST_PICK_THEME:
                subtitle = -1;
                title = String.format(formatString, theme.getTitle(), theme.getAuthor());
                break;
            case REQUEST_PICK_DAY_THEME:
                subtitle = R.string.pref_appearance_day_time_theme_title;
                title = String.format(formatString, dayTheme.getTitle(), dayTheme.getAuthor());
                break;
            case REQUEST_PICK_NIGHT_THEME:
                subtitle = R.string.pref_appearance_night_time_theme_title;
                title = String.format(formatString, nightTheme.getTitle(), nightTheme.getAuthor());
                break;
        }

        toolbar.setTitle(title);

        if (subtitle == -1)
            toolbar.setSubtitle("");
        else
            toolbar.setSubtitle(subtitle);
    }

    private void onRefresh() {
        getThemes();
    }

    private void getThemes() {
        ArrayList<ThemeItem> items = new ArrayList<>(ThemeEngine.themes);
        ArrayList<ThemeItem> deleteItems = new ArrayList<>();

        for (ThemeItem item : items) {
            if (request == REQUEST_PICK_DAY_THEME && item.isDark()) {
                deleteItems.add(item);
            } else if (request == REQUEST_PICK_NIGHT_THEME && !item.isDark()) {
                deleteItems.add(item);
            }
        }

        items.removeAll(deleteItems);

        createAdapter(items);

        refreshLayout.setRefreshing(false);
    }

    private void createAdapter(ArrayList<ThemeItem> items) {
        if (adapter == null) {
            adapter = new ThemeAdapter(this, items);
            adapter.setOnItemClickListener(this::onItemClick);
            list.setAdapter(adapter);
            return;
        }

        adapter.changeItems(items);
        adapter.notifyItemRangeChanged(0, adapter.getItemCount());
    }

    private void onItemClick(View v, int position) {
        ThemeItem item = adapter.getItem(position);

        if (ThemeEngine.isThemeCompatible(item)) {
            if (animating ||
                    (request == REQUEST_PICK_THEME && item.equals(theme)) ||
                    (request == REQUEST_PICK_DAY_THEME && item.equals(ThemeEngine.getDayTheme())) ||
                    (request == REQUEST_PICK_NIGHT_THEME && item.equals(ThemeEngine.getNightTheme())))
                return;

            if (request == REQUEST_PICK_THEME) {
                ThemeEngine.setCurrentTheme(item.getId());
            } else if (request == REQUEST_PICK_DAY_THEME) {
                ThemeEngine.setDayTheme(item.getId());
            } else if (request == REQUEST_PICK_NIGHT_THEME) {
                ThemeEngine.setNightTheme(item.getId());
            }

            applyTitle();

            fadeLayout();

            navigationIcon.setTint(Color.RED); //я не знаю почему, но работает лишь так
            toolbar.setNavigationIcon(navigationIcon);
        } else {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.warning);
            builder.setMessage(item.getEngineVersion() > ThemeEngine.ENGINE_VERSION ? R.string.warning_theme_new : R.string.warning_theme_old);
            builder.setPositiveButton(android.R.string.ok, null);
            builder.show();
        }

    }

    private void fadeLayout() {
        animating = true;

        rootView.setAlpha(0);
        rootView.animate().alpha(1).setDuration(500).withEndAction(() -> animating = false).start();
    }
}

File: app/src/main/java/ru/melod1n/schedule/activity/ThemesActivity.kt
package ru.melod1n.schedule.activity

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.activity_themes.*
import kotlinx.android.synthetic.main.recycler_view.*
import ru.melod1n.schedule.R
import ru.melod1n.schedule.adapter.ThemeAdapter
import ru.melod1n.schedule.base.BaseActivity
import ru.melod1n.schedule.base.BaseAdapter
import ru.melod1n.schedule.common.ThemeEngine
import ru.melod1n.schedule.common.ThemeEngine.currentTheme
import ru.melod1n.schedule.common.ThemeEngine.dayTheme
import ru.melod1n.schedule.common.ThemeEngine.nightTheme
import ru.melod1n.schedule.model.ThemeItem
import java.util.*

class ThemesActivity : BaseActivity(), BaseAdapter.OnItemClickListener {

    private var adapter: ThemeAdapter? = null

    private var request = 0

    private var animating = false

    //    private Drawable navigationIcon;
    private var rootView: View? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_themes)

        setSupportActionBar(toolbar)

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        request = intent.getIntExtra("request", -1)

        applyBackground()
        applyTitle()

        rootView = toolbar.rootView

//        navigationIcon = ContextCompat.getDrawable(this, R.drawable.ic_arrow_backward);

        toolbar.setNavigationOnClickListener { onBackPressed() }

        refreshLayout.setOnRefreshListener { onRefresh() }

        recyclerView.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)

        initThemes()

        if (savedInstanceState == null) {
            scrollToCurrentTheme()
        }
    }

    private fun scrollToCurrentTheme() {
        var index = 0

        for (i in 0 until adapter!!.itemCount) {
            val item = adapter!!.getItem(i)
            if (item == currentTheme) index = i
        }

        recyclerView.smoothScrollToPosition(index)
    }

    private fun applyTitle() {
        val theme = currentTheme
        val dayTheme = dayTheme
        val nightTheme = nightTheme
        val formatString = "%s – %s"
        val title: String
        val subtitle: Int

        when (request) {
            REQUEST_PICK_THEME -> {
                subtitle = -1
                title = String.format(formatString, theme!!.title, theme.author)
            }
            REQUEST_PICK_DAY_THEME -> {
                subtitle = R.string.pref_appearance_day_time_theme_title
                title = String.format(formatString, dayTheme!!.title, dayTheme.author)
            }
            REQUEST_PICK_NIGHT_THEME -> {
                subtitle = R.string.pref_appearance_night_time_theme_title
                title = String.format(formatString, nightTheme!!.title, nightTheme.author)
            }
            else -> {
                subtitle = -1
                title = String.format(formatString, theme!!.title, theme.author)
            }
        }

        toolbar.title = title
        if (subtitle == -1) toolbar.subtitle = "" else toolbar!!.setSubtitle(subtitle)
    }

    private fun onRefresh() {
        initThemes()
    }

    private fun initThemes() {
        val items = ArrayList(ThemeEngine.themes)

        val deleteItems = ArrayList<ThemeItem>()

        for (item in items) {
            if (request == REQUEST_PICK_DAY_THEME && item.isDark) {
                deleteItems.add(item)
            } else if (request == REQUEST_PICK_NIGHT_THEME && !item.isDark) {
                deleteItems.add(item)
            }
        }

        items.removeAll(deleteItems)

        createAdapter(items)

        refreshLayout!!.isRefreshing = false
    }

    private fun createAdapter(items: ArrayList<ThemeItem>) {
        if (adapter == null) {
            adapter = ThemeAdapter(this, items).apply {
                onItemClickListener = this@ThemesActivity
            }

            recyclerView.adapter = adapter
            return
        }

        adapter!!.setItems(items)
        adapter!!.notifyDataSetChanged()
    }

    override fun onItemClick(position: Int) {
        val item = adapter!!.getItem(position)

        if (ThemeEngine.isThemeCompatible(item)) {
            if (animating ||
                    request == REQUEST_PICK_THEME && item == theme ||
                    request == REQUEST_PICK_DAY_THEME && item == dayTheme ||
                    request == REQUEST_PICK_NIGHT_THEME && item == nightTheme) return

            when (request) {
                REQUEST_PICK_THEME -> {
                    ThemeEngine.setCurrentTheme(item.id)
                }
                REQUEST_PICK_DAY_THEME -> {
                    ThemeEngine. setDayTheme(item.id)
                }
                REQUEST_PICK_NIGHT_THEME -> {
                    ThemeEngine.setNightTheme(item.id)
                }
            }

            applyTitle()
            fadeLayout()
        } else {
            val builder = AlertDialog.Builder(this)
            builder.setTitle(R.string.warning)
            builder.setMessage(if (item.engineVersion > ThemeEngine.ENGINE_VERSION) R.string.warning_theme_new else R.string.warning_theme_old)
            builder.setPositiveButton(android.R.string.ok, null)
            builder.show()
        }
    }

    private fun fadeLayout() {
        animating = true
        rootView!!.alpha = 0f
        rootView!!.animate().alpha(1f).setDuration(500).withEndAction { animating = false }.start()
    }

    companion object {
        const val REQUEST_PICK_THEME = 0
        const val REQUEST_PICK_DAY_THEME = 1
        const val REQUEST_PICK_NIGHT_THEME = 2
    }
}